### PR TITLE
Const propagate tuples

### PR DESCRIPTION
**Pull request by wesleywiser**
r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/transform/const_prop.rs [hunk 1]
```diff
@@ -374,10 +383,59 @@ impl<'a, 'mir, 'tcx> ConstPropagator<'a, 'mir, 'tcx> {
...
+                            }.into())
+                        } else {
+                            let mplace = self.use_ecx(source_info, |this| {
+                                Ok(this.ecx.allocate(place_layout, MemoryKind::Stack))
+                            }).unwrap();
+
+                            for (i, operand) in operands.iter().enumerate() {
+                                let op = self.eval_operand(operand, source_info)?;
+                                let should_continue = self.use_ecx(source_info, |this| {
+                                    if let Ok(imm) = this.ecx.try_read_immediate(op)? {
```
Comment (oli-obk) [inline src/librustc_mir/transform/const_prop.rs]: There's no need to check for scalars. You can write the `op` directly to the `field` without having to know whether it was an immediate. This way you can create arbitrary tuple combinations like `(5, (6, 7, 8))` and not just `(5, 6, 7, 8)`.

**Review by oli-obk — CHANGES_REQUESTED** [review]
General note: we're getting very much into "duplicating the miri-engine"-territory. If you look at https://github.com/rust-lang/rust/blob/1cbd8a4d686d1411105f26cddf876c5994e69593/src/librustc_mir/interpret/step.rs#L183-L205 you'll see that it does almost what you're doing. I think before taking this too far we should start using the locals in the `ecx`'s stack frame and thus be able to just call methods on the `ecx` for operations like these.

"not const propagatable" can be expressed via `Undef` bits until we have a better way. Basically if a variable depends on a runtime value, instead of having it be `None` like it is right now, we'd make it `Undef`, any reads will propagate that, but we'll never write an `Undef` to MIR. Any operations that fail with "read undef bytes" will just be ignored like we ignore `None` right now.

Does that make sense? Oh, my intention was not to hijack this PR. Let's take the discussion to zulip. You can just fixup the PR with the inline comments and we can merge it, but further improvements of this complexity level should probably be blocked on using `InterpCx` more.

**Comment by wesleywiser** [comment]
Yes, you're right I'd forgotten about that feedback. I'll start a thread on Zulip but I'm feeling like we should close this in favor of that change.

**Comment by wesleywiser** [comment]
Closing this in favor of reusing `InterpCx` as @oli-obk mentioned.

**Pull request closed without merging**